Let logical_not support non-bool tensors.

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24379 Implement operator logical_and
* #23978 Let logical_xor support non-bool tensors.
* **#23916 Let logical_not support non-bool tensors.**
* #23860 Recommend logical_not() instead of bitwise_not() when applying sub and neg on bool tensors.
* #23847 Add logical_xor operator
* #23839 Add logical_not operator.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

xuhdev committed:
[WIP] Let logical_not support non-bool tensors.

xuhdev committed:
Update on "[WIP] Let logical_not support non-bool tensors."

xuhdev committed:
Update on "[WIP] Let logical_not support non-bool tensors."

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

xuhdev commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -57,21 +57,26 @@ Tensor& bitwise_not_out(Tensor& result, const Tensor& self) {
+  logical_not_inplace_stub(iter.device_type(), iter);
+#ifdef BUILD_NAMEDTENSOR
+  at::namedinference::propagate_names(self, self);
+#endif
```
@zou3519 I suppose `propagate_names` is not needed here. Is my assumption correct?

zou3519 replied:
Yeah, it's not necessary anymore because TensorIterator handles name propagation already.

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

gchanan commented on torch/_torch_docs.py:
```diff
@@ -2543,17 +2543,22 @@ def merge_dicts(*dicts):
-""")
+    >>> torch.logical_not(torch.tensor([0, 1, -10], dtype=torch.int8))
+    tensor([ True, False,  False])
+    >>> torch.logical_not(torch.tensor([0, 1, -10], dtype=torch.double))
```
nit: this is a bit of a strange example -- we should probably have decimal values.

gchanan commented on torch/_torch_docs.py:
```diff
@@ -2543,17 +2543,22 @@ def merge_dicts(*dicts):
 logical_not(input, out=None) -> Tensor
 
-Computes the element-wise logical NOT of the given input tensor. The input tensor must have the bool dtype.
+Computes the element-wise logical NOT of the given input tensor. The output tensor always have the bool dtype. If the
```
I don't think this is true -- at least it shouldn't be.  We should support other dtypes for out, we just use bool if no other ones are provided.

gchanan commented on torch/_tensor_docs.py:
```diff
@@ -513,7 +513,17 @@ def add_docstr_all(method, docstr):
 logical_not_() -> Tensor
 
-In-place version of :meth:`~Tensor.logical_not`
+In-place version of :meth:`~Tensor.logical_not`, except that it always returns a tensor with the same dtype instead of
```
I don't think we want to document this separately -- it's too difficult to maintain.  The alternative dtype should be covered by the `out` argument for the non-inplace method.

gchanan commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -57,25 +57,24 @@ Tensor& bitwise_not_out(Tensor& result, const Tensor& self) {
-              "The output tensor of logical_not must be a bool tensor.");
-  auto iter = TensorIterator::unary_op(result, self,
-    /*check_internal_overlap=*/true);
+  TORCH_CHECK(result.scalar_type() == kBool, "The output tensor of logical_not must be a bool tensor.");
```
so, I don't know if TensorIterator currently supports this, but in NumPy I can have any return type (as long it's the same device?) and it will just do the cast for me.  I don't understand why we would need different behavior for in-place vs not.

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)

xuhdev committed:
Update on "Let logical_not support non-bool tensors."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/23916

Differential Revision: [D16719300](https://our.internmc.facebook.com/intern/diff/D16719300)
